addons/src/ao_colortip.c

```diff
@@ -0,0 +1,397 @@
+
+	/* Calculate range */
+	start = pos;
+	if (start >= 7)
+	{
+		start -= 7;
+	}
+	else
+	{
+		start = 0;
+	}
+	end = pos + 7;
+	if (end > max)
+	{
+		end = max;
+	}
+
+	/* Get text in range and examine it */
+	subtext = sci_get_contents_range(doc->editor->sci, start, end);
+	if (subtext != NULL)
+	{
+		pos = pos - start;
+		color = contains_color_value(subtext, pos, 1);
+		g_free(subtext);
+	}
+
+	return color;
+}
+
+static gboolean on_editor_button_press_event(GtkWidget *widget, GdkEventButton *event,
+											 AoColorTip *colortip)
+{
+	if (event->button == 1)
+	{
+		if (event->type == GDK_2BUTTON_PRESS)
+		{
+			AoColorTipPrivate *priv = AO_COLORTIP_GET_PRIVATE(colortip);
+
+			if (!priv->enable_double_click_color_chooser)
+				return FALSE;
+
+			if (get_color_value_at_current_doc_position() != -1)
+			{
+				keybindings_send_command
+					(GEANY_KEY_GROUP_TOOLS, GEANY_KEYS_TOOLS_OPENCOLORCHOOSER);
+			}
+		}
+	}
+	return FALSE;
+}
+
+
+void ao_color_tip_editor_notify(AoColorTip *colortip, GeanyEditor *editor, SCNotification *nt)
+{
+	ScintillaObject *sci = editor->sci;
+	AoColorTipPrivate *priv = AO_COLORTIP_GET_PRIVATE(colortip);
+
+	if (!priv->enable_colortip)
+	{
+		/* Ignore all events if color tips are disabled in preferences */
+		return;
+	}
+
+	switch (nt->nmhdr.code)
+	{
+		case SCN_DWELLSTART:
+		{
+			gchar *subtext;
+			gint start, end, pos, max;
+
+			/* Is position valid? */
+			if (nt->position < 0)
+				break;
+
+			/* Calculate range */
+			start = nt->position;
+			if (start >= 7)
+			{
+				start -= 7;
+			}
+			else
+			{
+				start = 0;
+			}
+			pos = nt->position - start;
+			end = nt->position + 7;
+			max = SSM(sci, SCI_GETTEXTLENGTH, 0, 0);
+			if (end > max)
+			{
+				end = max;
+			}
+
+			/* Get text in range and examine it */
+			subtext = sci_get_contents_range(sci, start, end);
+			if (subtext != NULL)
+			{
+				gint color;
+				
+				color = contains_color_value (subtext, pos, 2);
+				if (color != -1)
+				{
+					SSM(sci, SCI_CALLTIPSETBACK, color, 0);
+					SSM(sci, SCI_CALLTIPSHOW, nt->position, (sptr_t)"    ");
+				}
+				g_free(subtext);
+			}
+		}
+			break;
+
+		case SCN_DWELLEND:
+			SSM(sci, SCI_CALLTIPCANCEL, 0, 0);
+			break;
+	}
+}
+
+
+static void connect_document_button_press_signal_handler(AoColorTip *colortip, GeanyDocument *document)
+{
+	g_return_if_fail(DOC_VALID(document));
+
+	plugin_signal_connect(
+		geany_plugin,
+		G_OBJECT(document->editor->sci),
+		"button-press-event",
+		FALSE,
+		G_CALLBACK(on_editor_button_press_event),
+		colortip);
+}
+
+
+static void connect_documents_button_press_signal_handler(AoColorTip *colortip)
+{
+	guint i = 0;
+	/* connect the button-press event for all open documents */
+	foreach_document(i)
+	{
+		connect_document_button_press_signal_handler(colortip, documents[i]);
+	}
+}
+
+
+static void ao_color_tip_finalize(GObject *object)
+{
+	g_return_if_fail(object != NULL);
+	g_return_if_fail(IS_AO_COLORTIP(object));
+
```

Comment:
The AddOns-Plugin is calling ```plugin_module_make_resident()``` so do I really need to handle that?

Reply:
I'm not familiar with Addons machinery, but I think he means when you disable this feature and this GObject instance is deleted, there will still be connections to the dangling instance pointer. You should be able to test by enabling the feature, opening a few documents, disabling the feature, and then triggering button press events on the editor widgets.

Reply:
As @codebrainz said it depends how addons actually work, if the AoColortip object lives as long as the plugin does it would be fine I guess.

Reply:
@codebrainz, @b4n: if I understood it right all object should live as long as the plugin. But I will test it to verify it and report the result (as soon as I got time for it).

Reply:
Shouldn't the object be deleted (or not created in the first place) if the feature is disabled?

Reply:
I tested what @codebrainz described above and nothing happens. So I assume that everything is fine. If not I would have expected a crash if an old/invalid pointer would still be used.

Then I read @b4n's first comment again: AoColortip is only destroyed in function ```plugin_cleanup()```, so it should be OK.

Reply:
I wonder if I should throw away the whole GObject stuff and pass the config values directly from the ```ao_info``` struct in ```addons.c```. I only used it because I saw it in ```ao_markword.c```. I think it's quite complicated stuff for just saving to values. Is there any reason for not removing it?